File: src/k6s/engine/locks.py
"""File lock manager for coordinating file access between agents."""

from datetime import datetime, timedelta
from typing import Any

from k6s.models.context import FileLock
from k6s.store.db import Database


class LockResult:
    """Result of a lock acquisition attempt."""

    def __init__(
        self,
        success: bool,
        lock: FileLock | None = None,
        reason: str | None = None,
    ):
        self.success = success
        self.lock = lock
        self.reason = reason

    def to_dict(self) -> dict[str, Any]:
        """Convert to dictionary for MCP response."""
        result: dict[str, Any] = {"success": self.success}
        if self.lock:
            result["lock_token"] = self.lock.path
            result["expires_at"] = (
                self.lock.expires_at.isoformat() if self.lock.expires_at else None
            )
        if self.reason:
            result["reason"] = self.reason
        return result


class FileLockManager:
    """Manage exclusive file locks for coordinating agent access.

    Prevents multiple agents from editing the same file simultaneously.
    Locks have optional expiration to prevent deadlocks.
    """

    DEFAULT_LOCK_DURATION_SECONDS = 300  # 5 minutes

    def __init__(self, db: Database, session_id: str):
        self.db = db
        self.session_id = session_id

    async def acquire(
        self,
        path: str,
        agent_id: str,
        duration_seconds: int | None = None,
    ) -> LockResult:
        """Attempt to acquire a lock on a file.

        Uses BEGIN IMMEDIATE to serialize concurrent lock attempts
        and prevent two agents from both acquiring the same lock.

        Args:
            path: The file path to lock.
            agent_id: The agent requesting the lock.
            duration_seconds: Lock duration (default: 5 minutes).

        Returns:
            LockResult indicating success or failure.
        """
        async with self.db.transaction() as conn:
            # BEGIN IMMEDIATE ensures exclusive write access from the start,
            # preventing TOCTOU races between the SELECT and INSERT.
            await conn.execute("BEGIN IMMEDIATE")

            # Check for existing lock within the transaction
            cursor = await conn.execute(
                "SELECT * FROM file_locks WHERE path = ? AND session_id = ?",
                (path, self.session_id),
            )
            row = await cursor.fetchone()
            existing = FileLock.from_db_row(dict(row)) if row else None

            if existing:
                if existing.is_expired:
                    await conn.execute(
                        "DELETE FROM file_locks WHERE path = ? AND session_id = ?",
                        (path, self.session_id),
                    )
                elif existing.agent_id != agent_id:
                    return LockResult(
                        success=False,
                        reason=f"File locked by agent {existing.agent_id}",
                    )
                else:
                    # Same agent already has the lock, extend it
                    duration = duration_seconds or self.DEFAULT_LOCK_DURATION_SECONDS
                    new_expires = datetime.utcnow() + timedelta(seconds=duration)
                    await conn.execute(
                        "UPDATE file_locks SET expires_at = ? WHERE path = ? AND session_id = ?",
                        (new_expires.isoformat(), path, self.session_id),
                    )
                    existing.expires_at = new_expires
                    return LockResult(success=True, lock=existing)

            # Create new lock
            duration = duration_seconds or self.DEFAULT_LOCK_DURATION_SECONDS
            expires_at = datetime.utcnow() + timedelta(seconds=duration)

            lock = FileLock(
                path=path,
                session_id=self.session_id,
                agent_id=agent_id,
                expires_at=expires_at,
            )

            data = lock.to_db_row()
            columns = ", ".join(data.keys())
            placeholders = ", ".join("?" for _ in data)
            await conn.execute(
                f"INSERT OR REPLACE INTO file_locks ({columns}) VALUES ({placeholders})",
                tuple(data.values()),
            )

            return LockResult(success=True, lock=lock)

    async def release(self, path: str, agent_id: str) -> LockResult:
        """Release a lock on a file.

        Args:
            path: The file path to unlock.
            agent_id: The agent releasing the lock.

        Returns:
            LockResult indicating success or failure.
        """
        existing = await self._get_lock(path)
        if existing is None:
            return LockResult(success=True, reason="Lock not found (already released)")

        if existing.agent_id != agent_id:
            return LockResult(
                success=False,
                reason=f"Lock held by different agent: {existing.agent_id}",
            )

        await self._release_lock(path)
        return LockResult(success=True)

    async def check(self, path: str) -> FileLock | None:
        """Check if a file is locked.

        Returns the lock if held (and not expired), None otherwise.
        """
        lock = await self._get_lock(path)
        if lock and lock.is_expired:
            await self._release_lock(path)
            return None
        return lock

    async def is_locked(self, path: str) -> bool:
        """Check if a file is currently locked."""
        return await self.check(path) is not None

    async def get_holder(self, path: str) -> str | None:
        """Get the agent ID holding a lock on a file."""
        lock = await self.check(path)
        return lock.agent_id if lock else None

    async def list_locks(self, agent_id: str | None = None) -> list[FileLock]:
        """List all active locks in this session.

        Args:
            agent_id: Optional filter by agent.

        Returns:
            List of active FileLock objects.
        """
        if agent_id:
            sql = """
                SELECT * FROM file_locks
                WHERE session_id = ? AND agent_id = ?
            """
            params = (self.session_id, agent_id)
        else:
            sql = """
                SELECT * FROM file_locks
                WHERE session_id = ?
            """
            params = (self.session_id,)

        rows = await self.db.fetch_all(sql, params)
        locks = [FileLock.from_db_row(row) for row in rows]

        # Filter out expired locks and clean them up
        active_locks = []
        for lock in locks:
            if lock.is_expired:
                await self._release_lock(lock.path)
            else:
                active_locks.append(lock)

        return active_locks

    async def release_all_for_agent(self, agent_id: str) -> int:
        """Release all locks held by an agent.

        Returns the number of locks released.
        """
        return await self.db.delete(
            "file_locks",
            "session_id = ? AND agent_id = ?",
            (self.session_id, agent_id),
        )

    async def release_all(self) -> int:
        """Release all locks in this session.

        Returns the number of locks released.
        """
        return await self.db.delete(
            "file_locks",
            "session_id = ?",
            (self.session_id,),
        )

    async def _get_lock(self, path: str) -> FileLock | None:
        """Get a lock by path."""
        row = await self.db.fetch_one(
            "SELECT * FROM file_locks WHERE path = ? AND session_id = ?",
            (path, self.session_id),
        )
        if row is None:
            return None
        return FileLock.from_db_row(row)

    async def _release_lock(self, path: str) -> None:
        """Release a lock by path."""
        await self.db.delete(
            "file_locks",
            "path = ? AND session_id = ?",
            (path, self.session_id),
        )

    async def _extend_lock(
        self,
        lock: FileLock,
        duration_seconds: int | None,
    ) -> LockResult:
        """Extend an existing lock's expiration."""
        duration = duration_seconds or self.DEFAULT_LOCK_DURATION_SECONDS
        new_expires = datetime.utcnow() + timedelta(seconds=duration)

        await self.db.update(
            "file_locks",
            {"expires_at": new_expires.isoformat()},
            "path = ? AND session_id = ?",
            (lock.path, self.session_id),
        )

        lock.expires_at = new_expires
        return LockResult(success=True, lock=lock)

File: src/engine/locks.ts
/**
 * File lock manager for coordinating file access between agents.
 *
 * better-sqlite3 transactions are naturally serialized (single-threaded + sync),
 * so no TOCTOU race is possible.
 */

import type { Db } from "../store/db.js";
import {
  type FileLock,
  fileLockFromDbRow,
  fileLockToDbRow,
  isLockExpired,
} from "../models/context.js";

export interface LockResult {
  success: boolean;
  lock?: FileLock;
  reason?: string;
}

export function lockResultToDict(r: LockResult): Record<string, unknown> {
  const result: Record<string, unknown> = { success: r.success };
  if (r.lock) {
    result.lock_token = r.lock.path;
    result.expires_at = r.lock.expiresAt;
  }
  if (r.reason) result.reason = r.reason;
  return result;
}

const DEFAULT_LOCK_DURATION_SECONDS = 300;

export class FileLockManager {
  constructor(
    private db: Db,
    private sessionId: string,
  ) {}

  acquire(
    lockPath: string,
    agentId: string,
    durationSeconds?: number,
  ): LockResult {
    return this.db.transaction(() => {
      const row = this.db.fetchOne(
        "SELECT * FROM file_locks WHERE path = ? AND session_id = ?",
        [lockPath, this.sessionId],
      );
      const existing = row ? fileLockFromDbRow(row) : null;

      if (existing) {
        if (isLockExpired(existing)) {
          this.db.delete("file_locks", "path = ? AND session_id = ?", [
            lockPath,
            this.sessionId,
          ]);
        } else if (existing.agentId !== agentId) {
          return {
            success: false,
            reason: `File locked by agent ${existing.agentId}`,
          };
        } else {
          // Same agent — extend
          const duration = durationSeconds ?? DEFAULT_LOCK_DURATION_SECONDS;
          const newExpires = new Date(
            Date.now() + duration * 1000,
          ).toISOString();
          this.db.update(
            "file_locks",
            { expires_at: newExpires },
            "path = ? AND session_id = ?",
            [lockPath, this.sessionId],
          );
          return {
            success: true,
            lock: { ...existing, expiresAt: newExpires },
          };
        }
      }

      const duration = durationSeconds ?? DEFAULT_LOCK_DURATION_SECONDS;
      const lock: FileLock = {
        path: lockPath,
        sessionId: this.sessionId,
        agentId,
        acquiredAt: new Date().toISOString(),
        expiresAt: new Date(Date.now() + duration * 1000).toISOString(),
      };

      this.db.insertOrReplace("file_locks", fileLockToDbRow(lock));
      return { success: true, lock };
    });
  }

  release(lockPath: string, agentId: string): LockResult {
    const existing = this.getLock(lockPath);
    if (!existing) {
      return { success: true, reason: "Lock not found (already released)" };
    }
    if (existing.agentId !== agentId) {
      return {
        success: false,
        reason: `Lock held by different agent: ${existing.agentId}`,
      };
    }
    this.deleteLock(lockPath);
    return { success: true };
  }

  check(lockPath: string): FileLock | null {
    const lock = this.getLock(lockPath);
    if (lock && isLockExpired(lock)) {
      this.deleteLock(lockPath);
      return null;
    }
    return lock;
  }

  isLocked(lockPath: string): boolean {
    return this.check(lockPath) !== null;
  }

  getHolder(lockPath: string): string | null {
    const lock = this.check(lockPath);
    return lock?.agentId ?? null;
  }

  listLocks(agentId?: string): FileLock[] {
    const rows = agentId
      ? this.db.fetchAll(
          "SELECT * FROM file_locks WHERE session_id = ? AND agent_id = ?",
          [this.sessionId, agentId],
        )
      : this.db.fetchAll("SELECT * FROM file_locks WHERE session_id = ?", [
          this.sessionId,
        ]);

    const locks = rows.map(fileLockFromDbRow);
    const active: FileLock[] = [];
    for (const lock of locks) {
      if (isLockExpired(lock)) {
        this.deleteLock(lock.path);
      } else {
        active.push(lock);
      }
    }
    return active;
  }

  releaseAllForAgent(agentId: string): number {
    return this.db.delete(
      "file_locks",
      "session_id = ? AND agent_id = ?",
      [this.sessionId, agentId],
    );
  }

  releaseAll(): number {
    return this.db.delete("file_locks", "session_id = ?", [this.sessionId]);
  }

  private getLock(lockPath: string): FileLock | null {
    const row = this.db.fetchOne(
      "SELECT * FROM file_locks WHERE path = ? AND session_id = ?",
      [lockPath, this.sessionId],
    );
    return row ? fileLockFromDbRow(row) : null;
  }

  private deleteLock(lockPath: string): void {
    this.db.delete("file_locks", "path = ? AND session_id = ?", [
      lockPath,
      this.sessionId,
    ]);
  }
}
